monoDriveIO/monodrive-client
# Replay state buffer

### Overview
This is the client side of https://github.com/monoDriveIO/simulator/pull/397 See this PR for details

### Changes
- Adds several distributed examples using the new architecture (and updates the previous one). These include imgui plotting.
- Adds a command queue (FrameBuffer) for sending server commands on control flows where such a thing is necessary like the state update. For example, the physics thread can queue a vehicle command and then move on to sample the state without blocking. In retrospect, knowing what I know now about boost asio, this could be removed if we restructed the client side server stuff (probably the next big refactor). This would be better handled with boost asio context.
- FrameBuffer is added, a 1D FIFO that is thread safe for data handoffs. Used all over the place on client and server.
- UpdateState command added which is now the default way for updating state on replicas and in replay.
- Pause, unsupported call added for debugging, customer request. Sending true will pause simulation, false will unpause.
- 

### Testing
1. Run the distributed_plot example and see the glorious frame rates with no jitter on cameras.
    - Notice I swapped the configuration order, the replicas configure and start before the physics. This is possible with current 
2. architecture and prevents missed frames of physics updating without replicas ready.
3. Run replay example and see the glorious frame rates with no jitter on sensors.
4. Run each sensor example (CI/CD here we can run?) and verify that all sensors work with new architecture, performant, no crashes.
5. Stop the simulator while the client is connected and both are running. Should stop the simulation without hangs or crashes. Client should complain that it lost connection. A clever programmer will realize that you could reconnect and continue where you left of instead of restarting everything.
